backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunDAORdbImpl.java

```diff
@@ -3056,6 +3079,185 @@ public ExperimentRun cloneExperimentRun(
     }
 
     desExperimentRunBuilder.clearOwner().setOwner(authService.getVertaIdFromUserInfo(userInfo));
-    return insertExperimentRun(projectDAO, desExperimentRunBuilder.build(), userInfo);
+    return insertExperimentRun(
+        projectDAO, artifactStoreDAO, desExperimentRunBuilder.build(), userInfo);
+  }
+
+  private ExperimentRun populateEnvironmentBasedOnExperimentRunArtifacts(
+      ArtifactStoreDAO artifactStoreDAO, ExperimentRun experimentRun) throws ModelDBException {
+    ExperimentRun.Builder experimentRunBuilder = experimentRun.toBuilder();
+    List<Artifact> experimentRunArtifacts = experimentRun.getArtifactsList();
+    if (experimentRunArtifacts.isEmpty()) {
+      return experimentRun;
+    } else {
+      PythonEnvironmentBlob.Builder pythonEnvironmentBuilder = PythonEnvironmentBlob.newBuilder();
+      for (Artifact artifact : experimentRunArtifacts) {
+        if (artifact.getKey().equals(ModelDBConstants.MODEL_API_JSON)) {
+          addVersionInPythonEnvironmentBlob(artifactStoreDAO, pythonEnvironmentBuilder, artifact);
+        } else if (artifact.getKey().equals(ModelDBConstants.REQUIREMENTS_TXT)) {
+          addRequirementsInPythonEnvironmentBlob(
+              artifactStoreDAO, pythonEnvironmentBuilder, artifact);
+        }
+      }
+
+      if (pythonEnvironmentBuilder.getConstraintsCount() > 0
+          || pythonEnvironmentBuilder.hasVersion()) {
+        EnvironmentBlob.Builder environmentBlobBuilder = EnvironmentBlob.newBuilder();
+        environmentBlobBuilder.setPython(pythonEnvironmentBuilder.build());
+        experimentRunBuilder.setEnvironment(environmentBlobBuilder.build());
+      }
+    }
+    return experimentRunBuilder.build();
+  }
+
+  private void addRequirementsInPythonEnvironmentBlob(
+      ArtifactStoreDAO artifactStoreDAO,
+      PythonEnvironmentBlob.Builder pythonEnvironmentBuilder,
+      Artifact artifact)
+      throws ModelDBException {
+    InputStream inputStream = artifactStoreDAO.downloadArtifact(artifact.getPath());
+
+    try (BufferedReader br = new BufferedReader(new InputStreamReader(inputStream))) {
+      String line;
+      while ((line = br.readLine()) != null) {
+        Pattern pattern = Pattern.compile(ModelDBConstants.VER_SPEC_PATTERN);
```

Comment:
We do not need to recompile the pattern every iteration.  This should only need to be done once.